File: sdks/javascript/src/operators.ts
/**
 * Condition operators for flag evaluation.
 * Ports all 16 operators from the Go backend (internal/evaluation/operators.go).
 */

/**
 * Evaluates whether an attribute value satisfies a condition.
 */
export function evaluateCondition(attributeValue: unknown, operator: string, conditionValue: unknown): boolean {
  switch (operator) {
    case 'equals':
      return toString(attributeValue) === toString(conditionValue)
    case 'not_equals':
      return toString(attributeValue) !== toString(conditionValue)
    case 'contains':
      return evalContains(attributeValue, conditionValue)
    case 'not_contains':
      return !evalContains(attributeValue, conditionValue)
    case 'starts_with':
      return toString(attributeValue).startsWith(toString(conditionValue))
    case 'ends_with':
      return toString(attributeValue).endsWith(toString(conditionValue))
    case 'greater_than': {
      const pair = toFloat64Pair(attributeValue, conditionValue)
      return pair !== null && pair[0] > pair[1]
    }
    case 'less_than': {
      const pair = toFloat64Pair(attributeValue, conditionValue)
      return pair !== null && pair[0] < pair[1]
    }
    case 'gte': {
      const pair = toFloat64Pair(attributeValue, conditionValue)
      return pair !== null && pair[0] >= pair[1]
    }
    case 'lte': {
      const pair = toFloat64Pair(attributeValue, conditionValue)
      return pair !== null && pair[0] <= pair[1]
    }
    case 'in':
      return evalIn(attributeValue, conditionValue)
    case 'not_in':
      return !evalIn(attributeValue, conditionValue)
    case 'exists':
      return attributeValue != null
    case 'not_exists':
      return attributeValue == null
    case 'matches': {
      const pattern = toString(conditionValue)
      try {
        return new RegExp(pattern).test(toString(attributeValue))
      } catch {
        return false
      }
    }
    default:
      return false
  }
}

/**
 * Converts any value to its string representation.
 * Matches Go's fmt.Sprintf("%v", v) behavior for common types.
 */
function toString(v: unknown): string {
  if (v == null) return ''
  return String(v)
}

/**
 * Attempts to convert a value to a number (float64 equivalent).
 * Returns null if conversion fails.
 */
function toFloat64(v: unknown): number | null {
  if (typeof v === 'number') {
    return isNaN(v) ? null : v
  }
  if (typeof v === 'string') {
    const n = Number(v)
    return isNaN(n) ? null : n
  }
  return null
}

/**
 * Converts both values to numbers. Returns null if either conversion fails.
 */
function toFloat64Pair(a: unknown, b: unknown): [number, number] | null {
  const fa = toFloat64(a)
  const fb = toFloat64(b)
  if (fa === null || fb === null) return null
  return [fa, fb]
}

/**
 * Checks if the attribute contains the condition value.
 * For arrays: checks if the array contains the value (string comparison).
 * For strings: checks substring containment.
 */
function evalContains(attributeValue: unknown, conditionValue: unknown): boolean {
  if (Array.isArray(attributeValue)) {
    const target = toString(conditionValue)
    for (const item of attributeValue) {
      if (toString(item) === target) return true
    }
    return false
  }
  return toString(attributeValue).includes(toString(conditionValue))
}

/**
 * Checks if the attribute value is in the condition list.
 * The condition value should be an array (for SDK-side evaluation, the API
 * stores in/not_in values as JSON-encoded string arrays that are parsed before
 * reaching this function).
 */
function evalIn(attributeValue: unknown, conditionValue: unknown): boolean {
  if (!Array.isArray(conditionValue)) return false
  const target = toString(attributeValue)
  for (const item of conditionValue) {
    if (toString(item) === target) return true
  }
  return false
}

File: sdks/go/operators.go
package togglerino

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// evaluateCondition checks if a context value satisfies a condition.
// The condition value is always a string (the definitions API serializes all values).
// For in/not_in operators, the value is a JSON-encoded array string like ["US","CA"].
func evaluateCondition(cond ConditionDefinition, contextValue any) bool {
	switch cond.Operator {
	case "equals":
		return toString(contextValue) == cond.Value
	case "not_equals":
		return toString(contextValue) != cond.Value
	case "contains":
		return evalContains(contextValue, cond.Value)
	case "not_contains":
		return !evalContains(contextValue, cond.Value)
	case "starts_with":
		return strings.HasPrefix(toString(contextValue), cond.Value)
	case "ends_with":
		return strings.HasSuffix(toString(contextValue), cond.Value)
	case "greater_than":
		a, b, ok := toFloat64Pair(contextValue, cond.Value)
		return ok && a > b
	case "less_than":
		a, b, ok := toFloat64Pair(contextValue, cond.Value)
		return ok && a < b
	case "gte":
		a, b, ok := toFloat64Pair(contextValue, cond.Value)
		return ok && a >= b
	case "lte":
		a, b, ok := toFloat64Pair(contextValue, cond.Value)
		return ok && a <= b
	case "in":
		return evalIn(contextValue, cond.Value)
	case "not_in":
		return !evalIn(contextValue, cond.Value)
	case "exists":
		return contextValue != nil
	case "not_exists":
		return contextValue == nil
	case "matches":
		matched, err := regexp.MatchString(cond.Value, toString(contextValue))
		return err == nil && matched
	default:
		return false
	}
}

// toString converts any value to its string representation.
func toString(v any) string {
	if v == nil {
		return ""
	}
	return fmt.Sprintf("%v", v)
}

// toFloat64 attempts to convert a value to float64.
func toFloat64(v any) (float64, bool) {
	switch n := v.(type) {
	case float64:
		return n, true
	case float32:
		return float64(n), true
	case int:
		return float64(n), true
	case int8:
		return float64(n), true
	case int16:
		return float64(n), true
	case int32:
		return float64(n), true
	case int64:
		return float64(n), true
	case uint:
		return float64(n), true
	case uint8:
		return float64(n), true
	case uint16:
		return float64(n), true
	case uint32:
		return float64(n), true
	case uint64:
		return float64(n), true
	case string:
		f, err := strconv.ParseFloat(n, 64)
		if err != nil {
			return 0, false
		}
		return f, true
	default:
		return 0, false
	}
}

// toFloat64Pair converts both values to float64.
func toFloat64Pair(a, b any) (float64, float64, bool) {
	fa, okA := toFloat64(a)
	fb, okB := toFloat64(b)
	return fa, fb, okA && okB
}

// evalContains checks if the attribute contains the condition value.
// For strings, it checks substring containment.
// For slices, it checks if the slice contains the value.
func evalContains(attributeValue any, conditionValue string) bool {
	if slice, ok := toSlice(attributeValue); ok {
		for _, item := range slice {
			if toString(item) == conditionValue {
				return true
			}
		}
		return false
	}
	return strings.Contains(toString(attributeValue), conditionValue)
}

// evalIn checks if the attribute value is in the condition list.
// The condition value is a JSON-encoded array string like ["US","CA"].
func evalIn(attributeValue any, conditionValue string) bool {
	var list []any
	if err := json.Unmarshal([]byte(conditionValue), &list); err != nil {
		return false
	}
	target := toString(attributeValue)
	for _, item := range list {
		if toString(item) == target {
			return true
		}
	}
	return false
}

// toSlice attempts to convert a value to []any.
func toSlice(v any) ([]any, bool) {
	switch s := v.(type) {
	case []any:
		return s, true
	case []string:
		result := make([]any, len(s))
		for i, item := range s {
			result[i] = item
		}
		return result, true
	default:
		return nil, false
	}
}
